compat_test.go

```diff
@@ -189,6 +164,484 @@ func (*suite) TestBind(c *gc.C) {
+		id:      "root-id",
+		caveats: []caveat{{
+			condition: "wonderful",
+		}, {
+			condition: "bob-is-great",
+			location:  "bob",
+			rootKey:   "bob-caveat-root-key",
+		}},
+	}, {
+		location: "bob",
+		rootKey:  "bob-caveat-root-key",
+		id:       "bob-is-great",
+		caveats: []caveat{{
+			condition: "splendid",
+		}, {
+			condition: "barbara-is-great",
+			location:  "barbara",
+			rootKey:   "barbara-caveat-root-key",
+		}},
+	}},
+	conditions: []conditionTest{{
+		conditions: map[string]bool{
+			"wonderful": true,
+			"splendid":  true,
+		},
+		expectErr: `cannot find discharge macaroon for caveat "barbara-is-great"`,
+	}},
+}, {
+	about:     "recursive third party caveats",
+	macaroons: recursiveThirdPartyCaveatMacaroons,
+	conditions: []conditionTest{{
+		conditions: map[string]bool{
+			"wonderful":   true,
+			"splendid":    true,
+			"high-fiving": true,
+			"spiffing":    true,
+		},
+	}, {
+		conditions: map[string]bool{
+			"wonderful":   true,
+			"splendid":    true,
+			"high-fiving": false,
+			"spiffing":    true,
+		},
+		expectErr: `condition "high-fiving" not met`,
+	}},
+}, {
+	about: "unused discharge",
+	macaroons: []macaroonSpec{{
+		rootKey: "root-key",
+		id:      "root-id",
+	}, {
+		rootKey: "other-key",
+		id:      "unused",
+	}},
+	conditions: []conditionTest{{
+		expectFailure: []mcompat.Implementation{mcompat.ImplLibMacaroons},
+		expectErr:     `discharge macaroon "unused" was not used`,
+	}},
+}}
+
+var recursiveThirdPartyCaveatMacaroons = []macaroonSpec{{
+	rootKey: "root-key",
+	id:      "root-id",
+	caveats: []caveat{{
+		condition: "wonderful",
+	}, {
+		condition: "bob-is-great",
+		location:  "bob",
+		rootKey:   "bob-caveat-root-key",
+	}, {
+		condition: "charlie-is-great",
+		location:  "charlie",
+		rootKey:   "charlie-caveat-root-key",
+	}},
+}, {
+	location: "bob",
+	rootKey:  "bob-caveat-root-key",
+	id:       "bob-is-great",
+	caveats: []caveat{{
+		condition: "splendid",
+	}, {
+		condition: "barbara-is-great",
+		location:  "barbara",
+		rootKey:   "barbara-caveat-root-key",
+	}},
+}, {
+	location: "charlie",
+	rootKey:  "charlie-caveat-root-key",
+	id:       "charlie-is-great",
+	caveats: []caveat{{
+		condition: "splendid",
+	}, {
+		condition: "celine-is-great",
+		location:  "celine",
+		rootKey:   "celine-caveat-root-key",
+	}},
+}, {
+	location: "barbara",
+	rootKey:  "barbara-caveat-root-key",
+	id:       "barbara-is-great",
+	caveats: []caveat{{
+		condition: "spiffing",
+	}, {
+		condition: "ben-is-great",
+		location:  "ben",
+		rootKey:   "ben-caveat-root-key",
+	}},
+}, {
+	location: "ben",
+	rootKey:  "ben-caveat-root-key",
+	id:       "ben-is-great",
+}, {
+	location: "celine",
+	rootKey:  "celine-caveat-root-key",
+	id:       "celine-is-great",
+	caveats: []caveat{{
+		condition: "high-fiving",
+	}},
+}}
+
+func (*suite) TestVerify(c *gc.C) {
+	for i, test := range verifyTests {
+		c.Logf("test %d: %s", i, test.about)
+
+		for _, impl := range mcompat.Implementations {
```

Comment:
It might be nice if there were a Js, Go and Libmacaroon test suite that embeds a common base test suite that runs all the cases against the interface. That way, we could more easily switch one of the implementations off (if we didn't have all the dependencies for one of them available in a given dev env, let's say).